src/cmds/restic/cmd_forget.go

```diff
@@ -83,32 +80,40 @@ func runForget(opts ForgetOptions, gopts GlobalOptions, args []string) error {
 		return err
 	}
 
-	// Process all snapshot IDs given as arguments.
-	if len(args) != 0 {
-		for _, s := range args {
-			// Parse argument as hex string.
-			if _, err := hex.DecodeString(s); err != nil {
-				Warnf("argument %q is not a snapshot ID, ignoring\n", s)
-				continue
-			}
-			id, err := restic.FindSnapshot(repo, s)
-			if err != nil {
-				Warnf("could not find a snapshot for ID %q, ignoring\n", s)
-				continue
-			}
+	// group by hostname and dirs
+	type key struct {
+		Hostname string
+		Paths    []string
+		Tags     []string
+	}
+	snapshotGroups := make(map[string]restic.Snapshots)
 
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	for sn := range FindFilteredSnapshots(ctx, repo, opts.Host, opts.Tags, opts.Paths, args) {
+		if len(args) > 0 {
+			// When explicit snapshots args are given, remove them immediately.
 			if !opts.DryRun {
-				h := restic.Handle{Type: restic.SnapshotFile, Name: id.String()}
-				err = repo.Backend().Remove(h)
-				if err != nil {
+				h := restic.Handle{Type: restic.SnapshotFile, Name: sn.ID().String()}
+				if err = repo.Backend().Remove(h); err != nil {
 					return err
 				}
-
-				Verbosef("removed snapshot %v\n", id.Str())
+				Verbosef("removed snapshot %v\n", sn.ID().Str())
 			} else {
-				Verbosef("would remove snapshot %v\n", id.Str())
+				Verbosef("would have removed snapshot %v\n", sn.ID().Str())
+			}
+		} else {
+			var tags []string
+			if opts.GroupByTags {
+				tags = sn.Tags
+				sort.StringSlice(tags).Sort()
 			}
+			sort.StringSlice(sn.Paths).Sort()
+			k, _ := json.Marshal(key{Hostname: sn.Hostname, Tags: sn.Tags, Paths: sn.Paths})
```

Comment:
Uh, you're throwing away an error here, that's not a good idea. I agree that there shouldn't be an error most of the time, but if there is one, `json.Marshal()` may return an empty buffer, so `string(buf)` will be the empty string, which means that all snapshots are sorted into the same group. That's not a good idea.